File: 39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#

# @lc code=start
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        self.checkCombination(candidates, target, 0, 0, [], res)
        return res

    def checkCombination(self, candidates, target, curr_idx, curr_sum, curr_combo, res):
        if (curr_sum > target):
            return
        elif (curr_sum == target):
            res.append(curr_combo[:])
            return
        
        for i in range(curr_idx, len(candidates)):
            curr_combo.append(candidates[i])
            curr_sum += candidates[i]
            self.checkCombination(candidates, target, i, curr_sum, curr_combo, res)
            curr_combo.pop()
            curr_sum -= candidates[i]
        
# @lc code=end

File: 39.combination-sum.java
/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        checkCombinations(candidates, target, 0, 0, new ArrayList<>(), res);
        return res;
    }

    private void checkCombinations(int[] candidates, int target, int currentIdx, int currentSum, List<Integer> currentCombo, List<List<Integer>> res) {
        
        if (currentSum == target) {
            res.add(new ArrayList<>(currentCombo));
        }

        for (int i = currentIdx; i < candidates.length; i++) {
            if (currentSum + candidates[i] > target) {
                continue;
            }
            currentCombo.add(candidates[i]);
            currentSum += candidates[i];
            checkCombinations(candidates, target, i, currentSum, currentCombo, res);
            currentCombo.remove(currentCombo.size() - 1);
            currentSum -= candidates[i];
        }
    }
}
// @lc code=end

